### PR TITLE
WIP refactoring/adding featureType checks

### PR DESCRIPTION
**Pull request by daltonkell**
### Motivation

Outstanding issue #845 and discussion #857 reveal the nuances and challenges faced when validating a data set's `featureType`. The CF Discrete Sampling Geometries provide many flexible options for the logical structure of storing discretely sampled data in a NetCDF file, but with flexibility comes complexity. This is a fairly large change in the logic behind checking featureTypes.

### What's changed?

Existing Compliance Checker logic checked orthogonal multidimensional array representations. This is the trivial way of storing data and the most easily conceptualized. The CF Discrete Sampling Geometries allow data to be stored in ragged array formats, but the Compliance Checker did not check the validity of such structures. Moreover, because of this, undesirable behavior was documented when running the compliance checker on data sets which in fact did conform to the standard (see #845).

This pull request addresses both the documented issue and the underlying source by adding methods to properly check for DSGs represented by ragged array storage structure. It also addresses the use of outdated features, such as "multi-timeseries-orthogonal" or "profile-incomplete", which are no longer specified in the CF specification. The methods which produced these results are not completely removed as their logic is still strong, but are instead incorporated into "wrapper methods" which output the parent feature instead of the sub-type.

Unit tests are included with the new methods and existing tests are adapted.

### Proof of concept

The undesired behavior in #845 is no longer exhibited:

```
(env38) dalton@ubuntu:~/compliance-checker$ python cchecker.py -t cf test_ragged_profile.nc                         
Running Compliance Checker on the datasets from: ['test_ragged_profile.nc']


--------------------------------------------------------------------------------
                         IOOS Compliance Checker Report                         
                          Version 4.3.4.dev79+g250f0b4                          
                     Report generated 2021-04-05T16:20:02Z                      
                                     cf:1.7                                     
http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html
--------------------------------------------------------------------------------
                               Corrective Actions                               
test_ragged_profile.nc has 9 potential issues


                                     Errors                                     
--------------------------------------------------------------------------------
§3.3 Standard Name
* Attribute long_name or/and standard_name is highly recommended for variable LONGITUDE
* Attribute long_name or/and standard_name is highly recommended for variable TIME
* Attribute long_name or/and standard_name is highly recommended for variable LATITUDE
* Attribute long_name or/and standard_name is highly recommended for variable TEMP
* Attribute long_name or/and standard_name is highly recommended for variable NOMINAL_DEPTH

§4.1 Latitude Coordinate
* latitude variable 'LATITUDE' must define units

§4.2 Longitude Coordinate
* longitude variable 'LONGITUDE' must define units

§4.3 Vertical Coordinate
* NOMINAL_DEPTH: vertical coordinates not defining pressure must include a positive attribute that is either 'up' or 'down'
* §4.3.1 NOMINAL_DEPTH's units must be defined for vertical coordinates, there is no default

§4.4 Time Coordinate
* TIME does not have units


                                    Warnings                                    
--------------------------------------------------------------------------------
§2.6 Attributes
* §2.6.2 global attribute title should exist and be a non-empty string
* §2.6.2 global attribute history should exist and be a non-empty string
* §2.6.1 Conventions field is not present

§3.1 Units
* units (None) attribute of 'LONGITUDE' must be a string compatible with UDUNITS
* units (None) attribute of 'TIME' must be a string compatible with UDUNITS
* units (None) attribute of 'LATITUDE' must be a string compatible with UDUNITS
* units (None) attribute of 'TEMP' must be a string compatible with UDUNITS
* units (None) attribute of 'NOMINAL_DEPTH' must be a string compatible with UDUNITS

§4.1 Latitude Coordinate
* latitude variable 'LATITUDE' should define valid units for latitude

§4.2 Longitude Coordinate
* longitude variable 'LONGITUDE' should define valid units for longitude
```

@mhidas
@benjwadams
@openSourcerer9000
@ocefpaf

This is a WIP, so all feedback is welcome. We can add more here or make modifications elsewhere.

### CODE REVIEW

File: compliance_checker/cf/cf.py [hunk 1]
```diff
@@ -4457,48 +4432,49 @@ def check_variable_features(self, ds):
...
         for name in self._find_geophysical_vars(ds):
             variable_feature = cfutil.guess_feature_type(ds, name)
             # If we can't figure it out, don't check it.
             if variable_feature is None:
                 continue
+            feature_types_found[variable_feature].append(name)            
             matching_feature = TestCtx(BaseCheck.MEDIUM, self.section_titles["9.1"])
             matching_feature.assert_true(
-                variable_feature in feature_type_map[feature_type],
+                variable_feature == _feature,
```
Comment (mhidas) [inline compliance_checker/cf/cf.py:4456]: As above, this should be a case-insensitive comparison.

**Review by mhidas — COMMENTED** [review]
Thanks @daltonkell for working on these issues. Sorry I'm a bit late to provide feedback. I haven't had time to do a proper code review, only had a very quick look.

I'm now testing the latest code on master against our files. The "Unidentifiable feature" errors (#850) are gone :+1: 
However, I now get warnings that are still incorrect, e.g. for [this file](https://s3-ap-southeast-2.amazonaws.com/imos-data/IMOS/SOOP/SOOP-XBT/DELAYED/Line_IX28_Dumont-d-Urville-Hobart/2019/IMOS_SOOP-XBT_T_20190108T073100Z_IX28_FV01_ID-89011149.nc) which certainly _is_ a `profile`:
```
                                    Warnings                                    
--------------------------------------------------------------------------------
§9.1 Features and feature types
* TEMP is not a profile, it is detected as a ragged-array
```
Also found similar warnings for [timeseries](https://s3-ap-southeast-2.amazonaws.com/imos-data/IMOS/SOOP/SOOP-TMV/VLST_Spirit-of-Tasmania-1/realtime/mooring/1sec/2019/IMOS_SOOP-TMV_TSUB_20190101T074611Z_VLST_FV00_mooring-DEV_END-20190101T101530Z.nc) and [trajectory](https://s3-ap-southeast-2.amazonaws.com/imos-data/IMOS/SOOP/SOOP-TMV/VLST_Spirit-of-Tasmania-1/realtime/transect/1sec/2019/IMOS_SOOP-TMV_TSUB_20190101T101531Z_VLST_FV00_transect-D2M_END-20190101T192740Z.nc) files, which are also mis-identified as `ragged-array`.

Finally, as noted in the in-line comments, there are also some cases where the feature type is correct, but fails the check only because the global attribute is written with capitals (e.g. "timeSeries" which is valid).

**Comment by benjwadams** [comment]
@daltonkell has left IOOS projects to work on other opportunities.  I will take a look at these.  Would you mind creating issues for these?

**Comment by mhidas** [comment]
Thanks @benjwadams . The case sensitivity was fixed by #875. I have opened #877 for the ragged-array issue.

File: compliance_checker/cfutil.py [hunk 2]
```diff
@@ -882,6 +882,161 @@ def coordinate_dimension_matrix(nc):
...
+    # feature_type as we'll have to search for one with profile_id
+    # regardless; if single feature type, cf_role must match that
+    # featureType
+    cf_role_vars = nc.get_variables_by_attributes(cf_role=lambda x: x is not None)
+    if not cf_role_vars or \
+        (len(cf_role_vars)>1 and not is_compound) or \
+        (len(cf_role_vars)>2 and is_compound):
+        return False
+    cf_role_var = nc.get_variables_by_attributes(cf_role="{}_id".format(ftype))[0]
+    if cf_role_var.cf_role.split("_id")[0].lower() != ftype:
```
[the remark this answers is not in the document]
Comment (benjwadams) [reply compliance_checker/cfutil.py:913]: Seems to be.

**Pull request merged**